lib/cosmos/interfaces/protocols/stream_protocol.rb

```diff
@@ -122,62 +93,17 @@ def read
 
     # Writes the packet data to the stream.
     #
-    # @param packet [Packet] Packet data to write to the stream
-    def write(packet)
-      @write_mutex.synchronize do
-        packet = pre_write_packet(packet)
-        data = packet.buffer(false)
-        if data
-          write_raw(data, false)
-          post_write_data(packet, data)
-        else
-          # write aborted - don't write data
-        end
-      end
-    end
-
-    # Writes the raw binary string to the stream.
-    #
-    # @param data [String] Raw binary string
-    # @param take_mutex [Boolean] Whether or not to take the write_mutex
-    def write_raw(data, take_mutex = true)
-      @write_mutex.lock if take_mutex
-      begin
-        if connected?()
-          @stream.write(pre_write_data(data))
-          @bytes_written += data.length
-        else
-          raise "Stream not connected for write_raw"
-        end
-      ensure
-        @write_mutex.unlock if take_mutex
-      end
-    end
-
-    protected
-
-    # Called to perform modifications on read data before making it into a packet
-    #
-    # @param packet_data [String] Raw packet data
-    # @return [String] Potentially modified packet data
-    def post_read_data(packet_data)
-      packet_data
-    end
-
-    # Called to perform modifications on a read packet before it is inserted
-    # into the current value table.
-    #
-    # @param packet [Packet] Original packet
-    # @return [Packet] Potentially modified packet
-    def post_read_packet(packet)
-      packet
+    # @param data [String] Packet data to write to the stream
+    def write_data(data)
+      @stream.write(data)
```

Comment:
The mutex and callbacks are all now handled by Interface so we simply write from the stream. It's a little weird to be depending on the stream variable to just be set by the higher level but it works. Perhaps this will break down in the other interface types: UDP, TCPServer, etc. Maybe stream should just be called 'link_layer' or something to indicate its the lowest level.